database/src/main/java/com/firebase/ui/database/FirebaseIndexListAdapter.java

```diff
@@ -3,57 +3,23 @@
- *     ListAdapter adapter = new FirebaseListAdapter<ChatMessage>(
- *             this,
- *             ChatMessage.class,
- *             android.R.layout.two_line_list_item,
- *             keyRef,
- *             dataRef)
- *     {
- *         protected void populateView(View view, ChatMessage chatMessage, int position)
- *         {
- *             ((TextView)view.findViewById(android.R.id.text1)).setText(chatMessage.getName());
- *             ((TextView)view.findViewById(android.R.id.text2)).setText(chatMessage.getMessage());
- *         }
- *     };
- *     listView.setListAdapter(adapter);
- * </pre>
- *
- * @param <T> The class type to use as a model for the data
- *           contained in the children of the given Firebase location
- */
 public abstract class FirebaseIndexListAdapter<T> extends FirebaseListAdapter<T> {
     /**
-     * @param activity    The activity containing the ListView
-     * @param modelClass  Firebase will marshall the data at a location into
-     *                    an instance of a class that you provide
-     * @param modelLayout This is the layout used to represent a single list item.
-     *                    You will be responsible for populating an instance of the corresponding
-     *                    view with the data from an instance of modelClass.
-     * @param keyRef      The Firebase location containing the list of keys to be found in {@code dataRef}.
-     *                    Can also be a slice of a location, using some
-     *                    combination of {@code limit()}, {@code startAt()}, and {@code endAt()}.
-     * @param dataRef     The Firebase location to watch for data changes.
-     *                    Each key key found in {@code keyRef}'s location represents
-     *                    a list item in the {@code ListView}.
+     * @param keyQuery The Firebase location containing the list of keys to be found in {@code
+     *                 dataRef}. Can also be a slice of a location, using some combination of {@code
+     *                 limit()}, {@code startAt()}, and {@code endAt()}.
+     * @param dataRef  The Firebase location to watch for data changes. Each key key found in {@code
+     *                 keyQuery}'s location represents a list item in the {@code ListView}.
+     * @see FirebaseListAdapter#FirebaseListAdapter(Activity, ObservableSnapshotArray, Class, int)
      */
     public FirebaseIndexListAdapter(Activity activity,
                                     Class<T> modelClass,
                                     @LayoutRes int modelLayout,
-                                    Query keyRef,
-                                    Query dataRef) {
-        super(activity, modelClass, modelLayout, new FirebaseIndexArray(keyRef, dataRef));
+                                    Query keyQuery,
+                                    DatabaseReference dataRef) {
+        init(activity, new FirebaseIndexArray<>(keyQuery, dataRef, this), modelClass, modelLayout);
```

Comment:
Here's a possible answer.  Since we're breaking parseSnapshot anyway, what if we removed it from the adapter entirely and instead had the adapters take an optional `SnapshotParser` in their constructor.  They would pass it through to the `FirebaseArray` if provided.

This seems to be a better representation of the hierarchy anyway,